File: ReLeetCode/0094.py
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        if root is None:
            return ans
        stk = []
        cur_node = root
        while cur_node is not None or len(stk) != 0: # cur_node is None说明当前这个树及其子树都遍历完了.
            while cur_node is not None:
                stk.append(cur_node)
                cur_node = cur_node.left
            cur_node = stk.pop()
            ans.append(cur_node.val)
            cur_node = cur_node.right
        return ans

# 迭代 模拟遍历二叉树过程.
# 从根节点出发, 沿着左子树一直走到叶节点, 经过的节点依次入栈. (1)
# 叶节点后, 栈顶元素出栈cur_node, 把元素放到结果数组中, 让cur_node指向cur_node右子树. (2)
# 对当前cur_node按照(1)过程循环处理. (3)

# 时间复杂度 O(n)
# 空间复杂度 O(n)


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        if root is None:
            return ans
        ans.extend(self.inorderTraversal(root.left))
        ans.append(root.val)
        ans.extend(self.inorderTraversal(root.right))
        return ans

# 递归

# 时间复杂度 O(n)
# 空间复杂度 O(lgn)

File: ReLeetCode/0094.cpp
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> ans;
        if (root == nullptr) {
            return ans;
        }
        stack<TreeNode*> stk;
        while (root != nullptr || !stk.empty()) {
            while (root) {
                stk.push(root);
                root = root->left;
            }
            auto mid = stk.top();
            stk.pop();
            ans.push_back(mid->val);
            root = mid->right;
        }
        return ans;
    }
};

// 时间复杂度 O(n)
// 空间复杂度 O(n)

// 使用 root 记录当前正在指向的节点，使用 stk 存储待遍历的节点